ref/src/sw_driver.F90

```diff
@@ -116,10 +155,35 @@ program sw_driver
     stop 1
   endif
 
+  ! Copy u to usave for testing the exchange
+    usave = u
 
   ! Get the start time
   call system_clock(count_start, count_rate)
 
+#ifdef ENABLE_MPI
+  ! Exchange the halo for the variable u
+#ifdef ENABLE_GPTL
+  if (do_profile == 1) then
+     ret = gptlstart('exchange')
+  end if
+#endif
+  call exchange
+#ifdef ENABLE_GPTL
+  if (do_profile == 1) then
+     ret = gptlstop('exchange')
+  end if
+#endif
+
+  !Test the exchange
+  call test_exchange(matches)
```

Comment:
It is important to do this test, but we should have the tests done separately from the kernel.

Reply:
@christopherwharrop-noaa  I agree we should do this test separately, but I think we should also do it here since takes little time and it's a standard check to make sure everything is as it should be.

Reply:
The issue isn't how expensive the test is, but how the kernel is organized and what it does.  I feel that the kernel code should only contain the implementation of the kernel computation and should not contain tests that check whether the computations work correctly.  You could/would not do that test in a real code that runs in production.  I feel that the kernel should just run the kernel and that is is the job of our test suite test to see if the code that does the halo exchange works correctly.  I think there needs to be a clear separation between the kernel computation and the kernel tests, particularly because this is the reference implementation that others will follow.  If we blend the the two together it makes it very confusing for others which things are the kernel and which things are the tests.

Reply:
Real codes have all kinds of error checks. In FIM and NIM MPI is encapsulated into SMS and SMS has sanity checks all over the place so it seems to me this error check is in the same sanity-check spirit.

Reply:
Agree that real codes have many kinds of error checks.  Disagree that this error check is in the spirit of sanity checks, particularly since (1) this check could not be done in a real code without extra communication cost, and (2) given a correct exchange algorithm, the only way this could fail is due to the network stack dropping bits.  This check is not a sanity check; it is a test to check whether the algorithm works.  Which is precisely what we need for our test suite, and which should also not be in the main body of the code.  I can not approve this PR until this test is removed from the driver and moved to the test suite.